// sqlit/dist/engine/mysql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require(".");
const mysql = require("mysql");
class _ConnectionPool extends _1.ConnectionPool {
    constructor(options) {
        super();
        this.pool = mysql.createPool(options);
    }
    getConnection() {
        return new Promise((resolve, reject) => {
            return this.pool.getConnection((error, connection) => {
                if (error)
                    reject(Error(error));
                resolve(new _Connection(connection, true));
            });
        });
    }
    end() {
        return new Promise((resolve, reject) => {
            return this.pool.end(error => {
                if (error)
                    reject(error);
                else
                    resolve();
            });
        });
    }
    escape(value) {
        return mysql.escape(value);
    }
    escapeId(name) {
        return mysql.escapeId(name);
    }
}
class _Connection extends _1.Connection {
    constructor(options, connected) {
        super();
        this.dialect = 'mysql';
        this.queryCounter = new _1.QueryCounter();
        if (connected) {
            this.connection = options;
        }
        else {
            this.connection = mysql.createConnection(options);
        }
    }
    release() {
        this.connection.release();
    }
    query(sql) {
        this.queryCounter.total++;
        return new Promise((resolve, reject) => this.connection.query(sql, (error, results, fields) => {
            if (error) {
                return reject(error);
            }
            if (results.insertId) {
                resolve(results.insertId);
            }
            else {
                resolve(results);
            }
        }));
    }
    transaction(callback) {
        return new Promise((resolve, reject) => {
            return this.connection.beginTransaction(error => {
                if (error)
                    return reject(error);
                let promise;
                try {
                    promise = callback(this);
                }
                catch (error) {
                    return this.connection.rollback(() => {
                        reject(error);
                    });
                }
                if (promise instanceof Promise) {
                    return promise
                        .then(result => this.connection.commit(error => {
                        if (error) {
                            return this.connection.rollback(() => {
                                reject(error);
                            });
                        }
                        else {
                            resolve(result);
                        }
                    }))
                        .catch(reason => this.connection.rollback(() => {
                        reject(reason);
                    }));
                }
                else {
                    resolve();
                }
            });
        });
    }
    commit() {
        return new Promise((resolve, reject) => {
            this.connection.commit(error => {
                if (error)
                    reject(error);
                else
                    resolve();
            });
        });
    }
    rollback() {
        return new Promise((resolve, reject) => {
            this.connection.rollback(error => {
                if (error)
                    reject(error);
                else
                    resolve();
            });
        });
    }
    end() {
        return new Promise((resolve, reject) => {
            this.connection.end(err => {
                if (err)
                    reject(err);
                else
                    resolve();
            });
        });
    }
    escape(value) {
        return mysql.escape(value);
    }
    escapeId(name) {
        return mysql.escapeId(name);
    }
}
exports.default = {
    createConnectionPool: (options) => {
        return new _ConnectionPool(options);
    },
    createConnection: (options) => {
        return new _Connection(options);
    }
};
